include/dlaf/cuda/executor.h

```diff
@@ -0,0 +1,220 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// @file
+
+#include <cstddef>
+#include <hpx/util/unwrap.hpp>
+#include <utility>
+#include <vector>
+
+#include <cublas_v2.h>
+#include <cuda_runtime.h>
+
+#include <hpx/include/async.hpp>
+#include <hpx/include/dataflow.hpp>
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/include/thread_executors.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/parallel/executors/execution_fwd.hpp>
+#include <hpx/util/yield_while.hpp>
+
+#include "dlaf/common/assert.h"
+#include "dlaf/cuda/error.h"
+#include "dlaf/cuda/pool.h"
+
+namespace dlaf {
+
+namespace internal {
+
+// Call the CUBLAS function and schedule an event after it to query when it's done.
+//
+// The function calls `cudaSetDevice()` multiple times on each thread, any previous assignment of CUDA
+// devices to threads is not preserved.
+template <typename F, typename... Ts>
+void run_cublas(int device, cublasHandle_t handle, cublasPointerMode_t pointer_mode, F f,
+                Ts... ts) noexcept {
+  // Set the device corresponding to the CUBLAS handle.
+  //
+  // The CUBLAS library context is tied to the current CUDA device [1]. A previous task scheduled on the
+  // same thread may have set a different device, this makes sure the correct device is used. The
+  // function is considered very low overhead call [2].
+  //
+  // [1]: https://docs.nvidia.com/cuda/cublas/index.html#cublascreate
+  // [2]: CUDA Runtime API, section 5.1 Device Management
+  DLAF_CUDA_CALL(cudaSetDevice(device));
+
+  // Use an event to query the CUBLAS kernel for completion. Timing is disabled for performance. [1]
+  //
+  // [1]: CUDA Runtime API, section 5.5 Event Management
+  cudaEvent_t event;
+  DLAF_CUDA_CALL(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+
+  // Get the stream on which the CUBLAS call is to execute.
+  cudaStream_t stream;
+  DLAF_CUBLAS_CALL(cublasGetStream(handle, &stream));
+
+  // Call the CUBLAS function `f` and schedule an event after it.
+  //
+  // The event indicates the the function `f` has completed. The handle may be shared by mutliple host
+  // threads, the mutex is here to make sure no other CUBLAS calls or events are scheduled between the
+  // call to `f` and it's corresponding event.
+  {
+    static hpx::lcos::local::mutex mt;
+    std::lock_guard<hpx::lcos::local::mutex> lk(mt);
+    DLAF_CUBLAS_CALL(cublasSetPointerMode(handle, pointer_mode));
+    DLAF_CUBLAS_CALL(f(handle, std::forward<Ts>(ts)...));
+    DLAF_CUDA_CALL(cudaEventRecord(event, stream));
```

Comment:
Yes, that would be better in principle. However, I'd rather not have the mutex at all by implementing the oversubscription approach I mention in the docs.

Reply:
"The main drawback of this approach is that it introduce some latency:
GPU finish computation -> CPU discover -> launch new kernel -> GPU execute kernel (compared to GPU finish computation -> GPU execute new kernel, if the new kernel is launched on the same stream or using events to sync the streams.)"

Note that in the hpx cuda_helper and cublas helper code, we provide both an `hpx::async(...)` call that inserts a callback (or in future an event) into the stream, and an `hpx::apply(...)` that implements the one-sided executor model where you fire and forget and do not get a future back from it. These can be mixed freely, so you can invoke N kernels using apply and then get a future back from the last one (or some other significant one), this allows you to avoid any cpu synchronization and leave it to the cuda stream management for kernels or cublas calls that don't need any cpu synchronization.

Anyway, I'm upgrading all the code to use an executor and adding @teonnik 's cuda event handling in too. I'll keep you updated with progress.